src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs

```diff
@@ -5349,5 +5349,296 @@ static void Main()
+00000000-0000-0000-0000-000000000000
+(True, 1c95cef0-1aae-4adb-a43c-54b2e7c083a0)
+(False, 00000000-0000-0000-0000-000000000000)
+(42, 8683f371-81b4-45f6-aaed-1c665b371594)
+(0, 00000000-0000-0000-0000-000000000000)
+(0, 00000000-0000-0000-0000-000000000000)
+(0, 00000000-0000-0000-0000-000000000000)";
+            var comp = CompileAndVerify(source, expectedOutput: expectedOutput);
+            comp.VerifyIL("Program.CoalesceInt32", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""int int?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceGeneric<T>", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""T T?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceTuple", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""(bool a, System.Guid b) (bool a, System.Guid b)?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceUserStruct", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""Program.S Program.S?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceStructWithImplicitConstructor", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""Program.S Program.S?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCoalesce_NullableWithConvertedDefault_Optimization()
+        {
+            var source = @"
+class Program
+{
+    static (bool a, System.Guid b, string c) CoalesceDifferentTupleNames((bool a, System.Guid b, string c)? x)
+    {
+        return x ?? default((bool c, System.Guid d, string e));
+    }
+
+    static void Main()
+    {
+        System.Console.WriteLine(CoalesceDifferentTupleNames((true, new System.Guid(""533d4d3b-5013-461e-ae9e-b98eb593d761""), ""value"")));
+        System.Console.WriteLine(CoalesceDifferentTupleNames(null));
+    }
+}";
+            var expectedOutput = 
+ @"(True, 533d4d3b-5013-461e-ae9e-b98eb593d761, value)
+(False, 00000000-0000-0000-0000-000000000000, )";
+            var comp = CompileAndVerify(source, expectedOutput: expectedOutput);
+            comp.VerifyIL("Program.CoalesceDifferentTupleNames", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""(bool a, System.Guid b, string c) (bool a, System.Guid b, string c)?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCoalesce_NullableWithNonDefault_NoOptimization()
+        {
+            var source = @"
+class Program
+{
+    static int? CoalesceWithNonDefault1(int? x)
+    {
+        return x ?? 2;
+    }
+
+    static int? CoalesceWithNonDefault2(int? x, int? y)
```

Comment:
>int? [](start = 48, length = 4)

Did you intend `y` to be an `int` rather than `int?`?

Reply:
Not necessarily, but I've added a new test case with `int? x, int y` parameters to make sure the case where right is a non-constant `T` is tested (same for VB).